src/omv/imlib/imlib.h

```diff
@@ -1136,24 +1136,33 @@ typedef struct imlib_draw_row_data {
     void *callback; // user
     void *callback_arg; // user
     void *dst_row_override; // user
-    int toggle; // private
-    void *row_buffer[2]; // private
-    #ifdef IMLIB_ENABLE_DMA2D
-    bool dma2d_request; // user
-    bool dma2d_enabled; // private
-    bool dma2d_initialized; // private
-    DMA2D_HandleTypeDef dma2d; // private
-    #if __DCACHE_PRESENT
-    void *dma2d_invalidate_addr; // private
-    int32_t dma2d_invalidate_dsize; // private
-    #endif
-    #endif
+    void *row_buffer; // private
     long smuad_alpha; // private
     uint32_t *smuad_alpha_palette; // private
 } imlib_draw_row_data_t;
 
 typedef void (*imlib_draw_row_callback_t) (int x_start, int x_end, int y_row, imlib_draw_row_data_t *data);
 
+typedef struct imlib_draw_gpu_data {
```

Comment:
* I think all of the GPU functions should be moved to `omv/common/omv_gpu.h` and named accordingly.
* Do we need `dst_x_end`, `dst_y_end`, `src_x_end`, `src_x_end` ? Aren't those deducible from x,y and `w` and `h` ? If not then pass as rectangles.
* Should use OR'd flags. `OMV_GPU_BILINEAR`, `OMV_GPU_HMIRROR` etc.. The good news is I can implement most of those. Probably also flags to use color/alpha ?
* What should black_background do ?
* LUTs will need to be portable. For example I need them to be `unsigned long`. They'll need to be defined in `omv_portconfig.h` as `omv_clut_t`, we can then update the generated LUTs to use that.

The `omv_gpu.h` should look like:

```C
int omv_gpu_draw_image(image_t *src,
                       point_t src_xy,
                       image_t *dst,
                       point_t dst_xy,
                       const omv_clut_t *alpha_palette,
                       const omv_clut_t *color_palette,
                       uint32_t int alpha,
                       uint32_t flags);
```

Reply:
> Do we need dst_x_end, dst_y_end, src_x_end, src_x_end ? Aren't those deducible from x,y and w and h ? If not then pass as rectangles.

Yeah, they can be rects. I didn't do that though since it produces an awkward amount of packing and unpacking.

> Actually no please pass the flags as those are useful. For example, I need to know if you want to scale or crop (crop when called from the sensor for set_window). The good news is I can support most of those flags, if not all of them.

Draw_image takes care of them already. The rects that are passed have already had all checks done on them. All the scaling and centring logic has been completely handled. I tried to write the logic so you could just directly pass the rects to the GPU blit method given it's docs.

> What should black_background do ?

In DMA2D is just sets the alpha for the background to 0 so that it acts as a black surface when blended.

> LUTs will need to be portable. For example I need them to be unsigned long. They'll need to be defined in omv_portconfig.h as omv_clut_t, we can then update the generated LUTs to use that.

The CLUTs are used directly. Code loads them into the hardware module. So, I don't think you need to make them portable?

> One more issue, it seems I'm allowed to do scaling in place, src==dst, it just works, so I don't need to allocate memory. Can we some how take advantage of that such that ? Maybe let gpu_draw_image allocate memory if it needs to ?

Yeah, I'll put some more cut-through paths in the draw image.

Reply:
* Yeah, they can be rects. I didn't do that though since it produces an awkward amount of packing and unpacking.

I think they'll just get passed on the stack, nothing extra, but will allow us to remove the extra gpu_data struct and just a call a function. Do we need the x/y ends ? Can't it infer those from the x/y start and w/h ?

> Draw_image takes care of them already

Understood, but I still need the flags because I need to know if you want to crop or scale. There's no way to know that other than passing yet another `bool`. At this point passing the flags is just easier, also covers future stuff.

> In DMA2D is just sets the alpha for the background to 0 so that it acts as a black surface when blended.

Can it be just an alpha arg then ?

> The CLUTs are used directly. Code loads them into the hardware module. So, I don't think you need to make them portable?

As I've said already I must load the CLUTs as ints, 32 bits per entry. It can't be uin8_t or 16 it must be 32 bits. Therefore it should be configured by each port.

Reply:
> As I've said already I must load the CLUTs as ints, 32 bits per entry. It can't be uin8_t or 16 it must be 32 bits. Therefore it should be configured by each port.

Yeah, so, for DMA2D there's a for loop that translates them into what the GPU needs:

```
if (!data->alpha_palette) {
            if (!data->color_palette) {
                for (int i = 0; i < 256; i++) {
                    clut[i] = (0xff << 24) | COLOR_Y_TO_RGB888(i);
                }
            } else {
                for (int i = 0; i < 256; i++) {
                    int pixel = data->color_palette[i];
                    clut[i] = (0xff << 24) |
                              (COLOR_RGB565_TO_R8(pixel) << 16) |
                              (COLOR_RGB565_TO_G8(pixel) << 8) |
                              COLOR_RGB565_TO_B8(pixel);
                }
            }
        } else {
            if (!data->color_palette) {
                for (int i = 0; i < 256; i++) {
                    clut[i] = (data->alpha_palette[i] << 24) | COLOR_Y_TO_RGB888(i);
                }
            } else {
                for (int i = 0; i < 256; i++) {
                    int pixel = data->color_palette[i];
                    clut[i] = (data->alpha_palette[i] << 24) |
                              (COLOR_RGB565_TO_R8(pixel) << 16) |
                              (COLOR_RGB565_TO_G8(pixel) << 8) |
                              COLOR_RGB565_TO_B8(pixel);
                }
            }
        }
```

This is hardware-specific. The LUTs passed in our library are designed for use in software. I can't change the type definition as that's the same as asking for all the software code to change that process them right now.

Reply:
I can, however, move this loop into draw_image(). I guess this would save it from being copied into each GPU driver.

I think it should be in the GPU driver, though, right?

Reply:
I see.. Well if nothing depends on their actual size, making them portable is just a `typedef` to make then as large as possible (i.e., 4 bytes).

Reply:
> I think it should be in the GPU driver, though, right as the LUT definition is hardware-specific.

Yes it is. The problem with doing them on the fly like that is that I have to generate them every frame which might be slow.

Reply:
Tested the driver and it takes it about 10 ms to blend an 720x480 image even with color table loading.